ocaml-lsp-server/src/ocaml_lsp_server.ml

```diff
@@ -353,41 +365,61 @@ let signature_help (state : State.t)
   let open Fiber.Result.O in
   let* doc = Fiber.return (Document_store.get store uri) in
   let pos = Position.logical position in
+  let client_capabilities = client_capabilities state in
   let prefix =
     (* The value of [short_path] doesn't make a difference to the final result
        because labels cannot include dots. However, a true value is slightly
        faster for getting the prefix. *)
     Compl.prefix_of_position (Document.source doc) pos ~short_path:true
   in
-  Document.with_pipeline_exn doc (fun pipeline ->
-      let typer = Mpipeline.typer_result pipeline in
-      let pos = Mpipeline.get_lexing_pos pipeline pos in
-      let node = Mtyper.node_at typer pos in
-      let context =
-        Merlin_analysis.Signature_help.application_signature node ~prefix
-      in
-      match context with
-      | `Application { fun_name; signature; parameters; active_param } ->
-        let fun_name = Option.value ~default:"_" fun_name in
-        let prefix = sprintf "%s : " fun_name in
-        let offset = String.length prefix in
-        let parameters =
-          List.map parameters
-            ~f:(fun (p : Merlin_analysis.Signature_help.parameter_info) ->
-              ParameterInformation.create
-                ~label:(`Offset (offset + p.param_start, offset + p.param_end))
-                ())
-        in
-        let label = prefix ^ signature in
-        let info = SignatureInformation.create ~label ~parameters () in
-        let help =
-          SignatureHelp.create ~signatures:[ info ] ~activeSignature:0
-            ?activeParameter:active_param ()
+  let open Fiber.O in
+  let* application_signature =
+    Document.with_pipeline_exn doc (fun pipeline ->
+        let typer = Mpipeline.typer_result pipeline in
+        let pos = Mpipeline.get_lexing_pos pipeline pos in
+        let node = Mtyper.node_at typer pos in
+        let signature =
+          Merlin_analysis.Signature_help.application_signature node ~prefix
         in
-        Ok (help, state)
-      | `Unknown ->
-        let help = SignatureHelp.create ~signatures:[] () in
-        Ok (help, state))
+        match signature with
+        | `Application a -> Some a
```

Comment:
I wonder if it makes sense for the merlin to just return an option type. In any case, we can change this we have signature help for things other than functions.